File: java/src/edu/wbl/basic/Employee.java
package edu.wbl.basic;

public class Employee {

    public int id;
    public double salary;
    public String name;

    public Employee() {

    }
    public Employee(int id) {
        this.id = id;
        //database
    }

    public String getDetails() {
        return name + ":" + salary;
    }
}

//Employee
//Manager is a Employee
//Director is a Manager
//Officer is a Director, Manager, Employee

//is a relationship
//Manager extends Employee = Sub Class extends Super Class
//Subclassing
//Manager will get everything from Employee except private variables and non default constructors.


//                                          Class               Variable            Method
//public  (Everywhere)                        Yes                   Yes                 Yes
//private (class)                             Yes(Inner class)      Yes                 Yes
//default (pkg)                               Yes                   Yes                 Yes
//protected (pkg + children outside)           No                    Yes                Yes


//public class Tesla {
//
//    private TComputer computer;
//
//    private class TComputer() {
//
//    }
//
//
//}

File: javascript/js-jan-2021/advanced/closure/employee.js
function getEmployee(id) {

    var empName = 'WBL';
    var id = id;
    var salary = 100.00;

    let getDetails = function() {
        return `${empName} - ${salary} - ${id}`;
    }

    let hikeSalary = function(pct) {
        salary += pct * 0.01 * salary;
    }

    return {
        "getDetails": getDetails,
        "hikeSalary": hikeSalary,
        name: 'Employee',
        "id": id
    }


}

let emp1 = getEmployee(10);
console.log(emp1.getDetails());
emp1.hikeSalary(5);
console.log(emp1.getDetails());